lib/integrations/awsoidc/aws_app_access_iam_config.go

```diff
@@ -0,0 +1,124 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package awsoidc
+
+import (
+	"context"
+	"log/slog"
+
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/iam"
+	"github.com/gravitational/trace"
+
+	awslib "github.com/gravitational/teleport/lib/cloud/aws"
+)
+
+const (
+	// defaultPolicyNameForAWSAppAccess is the default name for the Inline Policy added to the IntegrationRole.
+	defaultPolicyNameForAWSAppAccess = "AWSAppAccess"
+)
+
+// AWSAppAccessConfigureRequest is a request to configure the required Policies to use AWS App Access.
+type AWSAppAccessConfigureRequest struct {
+	// Region is the AWS Region.
+	// Used to set up the AWS SDK Client.
+	//Region string
+
+	// IntegrationRole is the Integration's AWS Role used to set up Teleport as an OIDC IdP.
+	IntegrationRole string
+
+	// IntegrationRoleAWSAppAccessPolicy is the Policy Name that is created to allow access to call AWS APIs.
+	// Defaults to AWSAppAccess
+	IntegrationRoleAWSAppAccessPolicy string
+}
+
+// CheckAndSetDefaults ensures the required fields are present.
+func (r *AWSAppAccessConfigureRequest) CheckAndSetDefaults() error {
+	if r.IntegrationRole == "" {
+		return trace.BadParameter("integration role is required")
+	}
+
+	if r.IntegrationRoleAWSAppAccessPolicy == "" {
+		r.IntegrationRoleAWSAppAccessPolicy = defaultPolicyNameForAWSAppAccess
+	}
+
+	return nil
+}
+
+// AWSAppAccessConfigureClient describes the required methods to create the IAM Policies required for AWS App Access.
+type AWSAppAccessConfigureClient interface {
+	// PutRolePolicy creates or replaces a Policy by its name in a IAM Role.
+	PutRolePolicy(ctx context.Context, params *iam.PutRolePolicyInput, optFns ...func(*iam.Options)) (*iam.PutRolePolicyOutput, error)
+}
+
+type defaultAWSAppAccessConfigureClient struct {
+	*iam.Client
+}
+
+// NewAWSAppAccessConfigureClient creates a new AWSAppAccessConfigureClient.
+func NewAWSAppAccessConfigureClient(ctx context.Context) (AWSAppAccessConfigureClient, error) {
+	cfg, err := config.LoadDefaultConfig(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if cfg.Region == "" {
+		// IAM Service does not support regions, however a value is required:
+		// https://github.com/aws/aws-sdk-go-v2/issues/1778#issuecomment-1210031692
+		// Providing an invalid region here, ensures the service uses the default AWS Partition.
+		cfg.Region = " "
+	}
+
+	return &defaultAWSAppAccessConfigureClient{
+		Client: iam.NewFromConfig(cfg),
+	}, nil
+}
+
+// ConfigureAWSAppAccess set ups the roles required for AWS App Access.
+// It creates an embedded policy with the following permissions:
+// - sts:AssumeRole
+//
+// The following actions must be allowed by the IAM Role assigned in the Client.
+//   - iam:PutRolePolicy
+func ConfigureAWSAppAccess(ctx context.Context, clt AWSAppAccessConfigureClient, req AWSAppAccessConfigureRequest) error {
```

Comment:
Why not make this a method in the "AWSAppAccessConfigureClient" instead?

Reply:
Could you please elaborate a little on this one?
I'm not sure I'm understanding what you mean.
`AWSAppAccessConfigureClient` is an interface, so adding another method means we would need to create another struct to have its implementation.

Reply:
I'm suggesting you add the ConfigureAWSAppAccess method directly to defaultAWSAppAccessConfigureClient, make it public, and ditch the AWSAppAccessConfigureClient interface completely.

Reply:
The goal of this type was to provide the IAMClient (actually only the PutRolePolicy method) for the default case.
Almost as a helper method that we use to get the desired client for production code.

If I'm making it public, than I think I will just call `iam.NewFromConfig(cfg)` in the `onIntegrationConfAWSAppAccessIAM` and pass it to `ConfigureAWSAppAccess`.


I don't like having a lot of code in `onIntegrationConfAWSAppAccessIAM`, so that's why I created the `NewAWSAppAccessConfigureClient`.
This wraps all AWS code into this package/file instead of spreading it to other files.

Reply:
Sorry, I'm a bit slow today.

> If I'm making it (defaultAWSAppAccessConfigureClient) public, than I think I will just call iam.NewFromConfig(cfg) in the onIntegrationConfAWSAppAccessIAM and pass it to ConfigureAWSAppAccess.

That seems reasonable to me. It doesn't look like NewAWSAppAccessConfigureClient is adding much value. Feel free to revert it to the initial commit where it returned AWSAppAccessConfigureClient, if you still prefer that one.